### ISSUE TITLE
build_runner watch stops detecting changes on Windows

### ISSUE THREAD

**Issue by dergemkr** [dart-lang/build#2161]
When I run the `pub run build_runner watch -o build` on Windows for an Angular project, it will stop detecting changes after one or two successful builds. This is not an issue on Linux - I've tested with the same project and Dart 2.2 on Linux and it continues to detect changes.

The command will usually get through one or two successful builds before it stops detecting changes, but interestingly, it will continue to detect an indefinite number of changes while the build continues to fail. Maybe it has something to do with writing the output directory? 

* Dart SDK Version: Dart VM version: 2.2.0 (Tue Feb 26 15:04:32 2019 +0100) on "windows_x64"
* Package Versions:
  * build: 1.1.0
  * build_runner: 1.2.3
  * build_web_compilers: 1.2.0

To ensure it wasn't specific to my project, I downloaded the [TOH-6 example](https://github.com/angular-examples/toh-6), upgraded build_web_compilers to 1.2.0and ran the above command on the project. I tested making changes to `web/main.dart` and `lib/app_component.dart` to see if changes were detected, and they were detected so long as the build failed.

**Comment by jakemac53** [comment]
On windows it uses a polling file watcher I believe (we just use package:watcher). It wouldn't surprise me if that combined with `-o` is really slow - but I wouldn't expect it to stop working entirely :/

**Comment by jakemac53** [comment]
Ok so actually it looks like we are explicitly choosing the polling watcher on windows https://github.com/dart-lang/build/blob/2fb938207512a4355a5d2e0af66a6bc46488dc08/build_runner/lib/src/generate/directory_watcher_factory.dart#L13 

I don't know if the linked issues are relevant still or not.

**Comment by dergemkr** [comment]
I suppose I should have clarified: By stops, I mean that it doesn't expose a crash message. There's just no subsequent builds after the build succeeded message. I just tested, and so long as "really slow" is not > 5 minutes, I don't think it's a delay.

I'm really liking this customizable build logic in Dart 2, though! I had always wanted this while working on Angular Dart < 4.

**Comment by jakemac53** [comment]
Interestingly it looks like we switched to a PollingDirectoryWatcher for exactly this reason (see https://github.com/dart-lang/build/issues/1031)?

I wonder if somehow you aren't picking up the polling directory watcher?

**Comment by dergemkr** [comment]
That issue does seem related - it continues to watch if I run with output directory `../build`. How do I determine if I'm using the polling watcher or not?

**Comment by jakemac53** [comment]
> it continues to watch if I run with output directory `../build`

That seems like it is the same issue for sure then.

One thing you could try is creating a dart script like this and running it:

```main.dart
import 'package:build_runner/src/generate/directory_watcher_factory.dart';

main() {
  print(defaultDirectoryWatcherFactory('.').runtimeType);
}
```

If that doesn't return a polling watcher then something is definitely weird.

If it does you could try snapshotting it first just in case and then running that `dart --snapshot=out.dill main.dart` followed by `dart out.dill`.

**Comment by dergemkr** [comment]
In both cases, I receive `PollingDirectoryWatcher` - I won't have time until this evening, but I could try checking into build_runner to see if it's using a different one.

**Comment by jakemac53** [comment]
@dergemkr are you still hitting this issue? were you able to get any more information?

**Comment by dergemkr** [comment]
I upgraded all dependencies, excluded the tests for toh-6, and retried the issue - the problem still exists. Changes keep triggering rebuilds if you target output directory "../build", but not after the first successful build if you target "build".

I added a print statement to get the watcher type of `PackageNodeWatcher._watcher` at created at line https://github.com/dart-lang/build/blob/master/build_runner/lib/src/watcher/node_watcher.dart#L36, and `WindowsDirectoryWatcher` was the output. I would have expected it to output `PollingDirectoryWatcher`, but I'm not very familiar with this package, though, so is this the right watcher to examine?

Dependency versions:
* Dart SDK: 2.3.1
* build: 1.1.4
* build_runner: 1.5.1
* build_web_compilers: 2.1.0

Thanks for the help!

lib\src\watcher\node_watcher.dart:36:
print("TAG: _watcher.runtimeType: ${_watcher.runtimeType}");
TAG: _watcher.runtimeType: WindowsDirectoryWatcher

Thanks for checking back!
